### PR TITLE
fix: meta changes (non-default) also update metaHOC state

### LINKED ISSUE #6492: [Bug]The File picker button is retained with the "1 file selected " text when the user tries to select the second file
## Description

[What happened] The File picker button is retained with the "1 file selected " text when the user tries to select the second file

### Steps to reproduce the behavior:
1.Click on Generate app 
2.Select s3 datasource 
3.Select bucket as "Assets-test.appsmith.com"
4.click on generate app button
5.Now click on Add file button 
6.Select one file 
7.click on the upload
8.now click on add file again 
9.verify the display

[LOOM DEMO](https://www.loom.com/share/c233efa813404e54b531d5a93e801546)

### Important Details

- Version: [Cloud ]
- OS: [.MacOS]
- Browser [chrome]
- Environment [ deploy preview->6264]

### LINKED ISSUE #9603: [Bug]: Input field is marked as error after form reset
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When user enables the "Reset Form on Success" option, the required input fields are marked as error.

### Steps To Reproduce

1. Create a form
2. Add an input field and mark it as required
3. Go to the property pane of the submit button and make sure "Reset Form on Success" is enabled
4. Deploy the app
5. Write something in the input field and click the submit button
6. After submit, the input field has red border for err
Screenshot 2021-12-06 at 5 55 22 PM
o
Screenshot 2021-12-06 at 5 54 18 PM
r

### Environment

Production

### Version

Cloud

### LINKED ISSUE #11350: [Bug]: resetWidget does not work for FilePicker widget
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

`resetWidget` fails to clear selected files in the file picker.

### Steps To Reproduce

Select a file and try to clear the widget.

### Environment

Production

### Version

Cloud

### PR DESCRIPTION
**Pull request by Rishabh-Rathod**
#### Problem
- When meta value changes from `isDirty = true` to `isDirty = false` on reset of inputWidget.
Previous condition to update state compared metaValue to be equal to default value which in such cases aren't defined.
#### Solution
- In metaHOC, we check if widget.meta has become empty => that only happens when we reset a widget, hence we initilise metaHOC state.

Fixes #9603
Fixes #11350
Fixes #6492

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Old test coverage

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>

    // Code coverage diff between base branch:release and head branch: fix/reset-metaHOC-staleState-9603 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | app/client/src/widgets/MetaHOC.tsx | 84.38 **(-0.87)** | 53.85 **(-4.04)** | 92.31 **(0)** | 83.05 **(-0.88)**
 :red_circle: | app/client/src/widgets/InputWidgetV2/widget/index.tsx | 73.6 **(-0.17)** | 60.87 **(-2.87)** | 81.25 **(0)** | 72.95 **(-0.16)**</details>

### CODE REVIEW

**Comment by Rishabh-Rathod** [comment]
ok-to-test sha=3362bac

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/LayoutWidgets/Tab_new_scenario_spec.js [hunk 1]
```diff
@@ -29,9 +25,10 @@ describe("Tab widget test", function() {
       .click({ force: true });
     cy.get(publish.checkboxWidget).should("be.visible");
     cy.get(publish.radioWidget).should("be.visible");
-    cy.get(publish.buttonWidget)
-      .contains("Confirm")
-      .click({ force: true });
+
+    cy.get(".t--widget-buttonwidget:contains('Confirm')").click({
```
Comment (hetunandu) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/LayoutWidgets/Tab_new_scenario_spec.js:29]: Why was this change in selector needed?

File: app/client/cypress/support/commands.js [hunk 2]
```diff
@@ -1416,7 +1416,7 @@ Cypress.Commands.add("PublishtheApp", () => {
   cy.route("POST", "/api/v1/applications/publish/*").as("publishApp");
   // Wait before publish
   // eslint-disable-next-line cypress/no-unnecessary-waiting
-  cy.wait(2000);
+  cy.wait(3000);
```
Comment (hetunandu) [inline app/client/cypress/support/commands.js:1419]: Another increase in wait. This one affects a lot of tests

File: app/client/src/widgets/MetaHOC.tsx [hunk 3]
```diff
@@ -57,27 +59,40 @@ const withMeta = (WrappedWidget: typeof BaseWidget) => {
...
+        );
+
+        const defaultPropertyExist = !!defaultProperty;
+        if (isMetaPropertyChanged) {
+          if (defaultPropertyExist) {
+            const isMetaEqualToDefault = _.isEqual(
+              this.props[defaultProperty],
+              this.props[metaProperty],
+            );
+            if (isMetaEqualToDefault) {
```
Comment (sbalaji1192) [inline app/client/src/widgets/MetaHOC.tsx:83]: We need to check here that state and this.props[metaProperty] is also not equal.

File: app/client/cypress/support/commands.js (hunk 2 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/support/commands.js]: undoing this change, and focusing on debugging the performance issue.

File: app/client/src/widgets/InputWidgetV2/widget/index.tsx [hunk 4]
```diff
@@ -393,8 +393,13 @@ class InputWidget extends BaseInputWidget<InputWidgetProps, WidgetState> {
 
   getPageView() {
     const value = this.props.text ?? "";
-    let isInvalid =
-      "isValid" in this.props && !this.props.isValid && !!this.props.isDirty;
+    let isInvalid = false;
```
[the remark this answers is not in the document]
Comment (sbalaji1192) [reply app/client/src/widgets/InputWidgetV2/widget/index.tsx:396]: Sounds good. Lets keep it.

**Review by sbalaji1192 — APPROVED** [review]

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/LayoutWidgets/Tab_new_scenario_spec.js (hunk 1 above)
Comment (Rishabh-Rathod) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/LayoutWidgets/Tab_new_scenario_spec.js]: This change is not needed and has been undone, it was added temporarily in an attempt to fix the failing cypress test.
Later realized the logic had an issue and the cypress test wasn't flaky.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Binding/Bind_TabWidget_Input_spec.js [hunk 5]
```diff
@@ -27,15 +27,18 @@ describe("Binding the input Widget with tab Widget", function() {
     cy.get(publish.tabWidget)
       .contains("Tab 2")
       .click({ force: true })
-      .should("be.selected");
+      .should("be.selected")
+      .wait(200);
```
[the remark this answers is not in the document]
Comment (Rishabh-Rathod) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Binding/Bind_TabWidget_Input_spec.js]: This changes are undone.

**Comment by Rishabh-Rathod** [comment]
Not sure why the previous review is not stale, please review it again.

**Review by hetunandu — APPROVED** [review]

**Review by sbalaji1192 — APPROVED** [review]

**Pull request merged**